valhalla/configuration.h

```diff
@@ -0,0 +1,57 @@
+#ifndef VALHALLA_CONFIGURATION_H_
+#define VALHALLA_CONFIGURATION_H_
+
+#include "baldr/rapidjson_utils.h"
+#include "filesystem.h"
+#include <boost/property_tree/ptree.hpp>
+#include <stdexcept>
+#include <string>
+
+namespace valhalla {
+namespace configuration {
+
+class Configuration {
+protected:
+  Configuration() = delete;
+  Configuration(const std::string& config_file_or_inline) {
+    if (config_file_or_inline.empty()) {
+      throw std::runtime_error("no config provided");
+    }
+
+    if (filesystem::is_regular_file(config_file_or_inline)) {
+      rapidjson::read_json(config_file_or_inline, config_);
+    } else {
+      auto inline_config = std::stringstream(config_file_or_inline);
+      rapidjson::read_json(inline_config, config_);
+    }
+  }
+
+  boost::property_tree::ptree config_;
+
+public:
+  Configuration(Configuration const&) = delete;
+  void operator=(const Configuration&) = delete;
+
+  static Configuration& instance(const std::string& config_file_or_inline = "") {
+    static Configuration instance(config_file_or_inline);
+    return instance;
+  }
```

Comment:
IMO that's a bit convoluted here. I also might not fully be aware of all constraints, but I'd prefer if this would immediately return the `config_`. I think you could remove the 2 free functions then, and the first call to this would need to specify a valid path to a valhalla.json or it'll throw in the constructor as it currently does. I think it's basically the same as your concept, just with a bit less code/fewer functions and all encapsulated in the class. Also, free functions with such generic names could be problematic at some point. You could call the function maybe rather `get_config` or so, it wouldn't be an instance anymore at that point. The important thing is that both function & `Configuration` are static. We could also return a `Configuration` instance from another function if we'd want to add more useful functions to `Configuration` class, which might be interesting.

Does that make sense?

Reply:
Another nit might be to call the class `config_singleton_t` instead of `Configuration`, following the existing pattern around singletons.

Reply:
I think I have an idea what you mean, I think it makes sense from an encapsulation point of view, will update.

Reply:
FYI, was just following up with @kevinkresier last week and he’d prefer to have a free function in the valhalla namespace, behaving similarly as I described above, so that we can just do valhalla::config(<optional_path>).

Reply:
I have adjusted the code a bit and taken out the configure-function, is it like you imagined?

Reply:
Jep, thanks a lot!